### DISCUSSION TITLE
Ideas for implementation of an event-driven LED system

### DISCUSSION THREAD

**Discussion by CapnHawke** [OpenStickCommunity/GP2040-CE#869] [Ideas]
In discussions on Discord across multiple servers, a user brought up a question, specifically, "can I set the LED to turn on after my button has been held for a set number of frames?" The ability to change an LED's behavior after a set period of time does not currently exist in the firmware. This feature request discusses potential ways to expand a user's ability to personally configure a wide set of variables in LED behavior, including specifying how an LED will behave while a corresponding button is pressed and held, and specifying how an LED will behave once the corresponding button has been held for a period of time defined by the user in the web configuration tool. 

**### Description of a potential solution:** 

It might be neat to be able to create an "advanced LED configuration" tool, enabling users to manually program and independently adjust sets of parameters as follows:

1. Giving the user the ability to tie LED behavior to four different states: (a) button not pressed; (b) button pressed and held; (c) button released; and/or (d) button pressed and held for X length of time, where X is decided by the user, and where execution of this instruction takes priority over LED behavior when a button is simply pressed and held.

2. Giving the user the ability to dictate a set of LED behaviors in each of these states. For example causing the LED to gently pulse, flash twice per second (or more rapidly if desired), gradually change color, or gradually fade from one brightness setting (e.g. 0) to another (e.g. 255) over a user defined length of time.

For example, with all of these features combined, if a user wishes, they could set an input timer for 3000ms, instruct the LED to blink amber twice per second while the timer is active, and then when the timer has completely run, instruct the LED to turn solid green. Alternatively, a player could instruct an LED to flash 5, 12, or 20 times per second while a button is being held. 

**### Other considerations:** 

Additionally, some users, particularly those with vision impairment, colorblindness, heightened sensitivity to peripheral lighting, etc., might be able to take advantage of manually configurable event-driven LED systems to help them improve their play by giving them additional visual cues for when they should begin to press or release a button. 

I have not yet done much research on how other systems implement their LEDs but I have seen other users asking about iCue implementation. iCue and Pico Fight Lights could be sources of inspiration for end goals with event driven LED implementation.

**Comment by TheTrainGoes** [comment]
Hi everyone,

This is general notice that we are closing down the discussion section of the GP2040-CE repo.  

If you would like to discuss anything further please open an issue or join our Discord:
https://discord.com/servers/openstickcommunity-1049366310389289001
